### DISCUSSION TITLE
[Closed] What's left until an official Meteor 3.0?

### DISCUSSION THREAD

**Discussion by denihs** [meteor/meteor#12865] [Announcements]
Our work to remove Fibers from Meteor began in 2021, which you can read about [here](https://github.com/meteor/meteor/discussions/11505). We’ve made significant progress and are now close to releasing an official version of Meteor 3.0 without Fibers and running on Node.js 20.

For those new to this, Fibers was our way of handling `promises`, but it’s not compatible with Node.js 16 and onwards. We're moving to `async` and `await` to match Node.js standards and community practices.

For background on Fibers and this change, check the [discussion](https://github.com/meteor/meteor/discussions/11505).

We’re in the last phase of this transition but still have much to do. This document outlines the remaining steps and will be continuously updated to keep the community informed.

### Estimated Release Dates

- Last Meteor 3.0 Beta:  March 26, 2024
- First Meteor 3.0 RC:  April 19, 2024
- Next RCs and Official: We need your help testing and providing feedback.

---

### Meteor Releases

<details>
  <summary><b>Meteor 3.0 Alpha (Finished)</b></summary>

- [x] [Unhandled Rejection when fails to bundle the app](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=41843545)  [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/12875)]
  - If the app server breaks, instead of waiting for changes, an error is thrown, and the app goes down.
- [x] [Review `Meteor.call`, `Meteor.callAsync`, `Meteor.apply`, `Meteor.applyAsync`](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872303) [[radekmie](https://github.com/radekmie), [denihs](https://github.com/denihs)]
  - We should review these methods to understand what's the best use for them. Today, Meteor.callAsync is the biggest problem. We solve most of these problems with some flags. But this isn't optimal, and we must find a good solution. For more context on this, we have this PR and this discussion.
- [x] Update Node to version 20.  [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/12866)]

</details>

<details>
  <summary><b>Meteor 3.0 Beta (Finished)</b></summary>

- [x] [Circle CI tests failing for external packages [Group 3]](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872498). [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/13065)]
  - Some tests are still failing in this Group. 
- [x] [Circle CI tests failing for external packages [Group 5]](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872508). [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/13067)]
  - Some tests are still failing in this Group.
- [x] Error with Source-map showing up in some cases. [Context](https://meteor-community.slack.com/archives/C01SWPPE81F/p1702314166182529). 
- [x] Insert not working for collections created of the fly. [Context](https://meteor-community.slack.com/archives/CN350MY1G/p1704695310758029) [[nachocodoner](https://github.com/nachocodoner)].
- [x] Fix Async iterators for Meteor 3 [context](https://forums.meteor.com/t/symbol-asynciterator-for-the-cursor-in-version-3-0-alpha-11/60549)
- [x] Review `observeChanges`  -  [#12918 [Meteor 3] `Mongo.Cursor.observeChanges` returns a Promise instead of a query handle](https://github.com/meteor/meteor/issues/12918) [[nachocodoner](https://github.com/nachocodoner)][[PR](https://github.com/meteor/meteor/pull/13024)]
- [x] [accounts-password: Meteor.user() doesn't return the user inside a Meteor.createUser() anymore](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872647). [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/13027)]
  - Some tests heavily depend on the behavior, like [this one].(https://github.com/meteor/meteor/blob/ce22f2d548d237b6a1d70064a86b106a853021cc/packages/accounts-password/password_tests.js#L132-L134). [This](https://github.com/meteor/meteor/blob/ce22f2d548d237b6a1d70064a86b106a853021cc/packages/accounts-base/accounts_client.js#L370-L372) was the workaround used to make the tests pass. It's not the ideal solution. We need to understand why this was happening and fix it.
 - [x] [Fix minifying lazy stylesheets](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872533).
  - It is required for Tailwind to work. The minifier is given the content as a promise instead of as a string, which causes the `custom minifier - devel vs prod` test to fail.
- [x] [Fix Linker bug](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872545) [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/13031)]
  - In this [commit](https://github.com/meteor/meteor/pull/12580/commits/eb63a1b6643805590d3bf70ff9611e837aede1e9#diff-0972e87d5bf49b0ff289a29dee4ae83bb04b83d2e057976f7a7680e0e1ee2e0aR176-R177), we introduced a hack to avoid returning the same file more than once in the function getPrelinkedFiles().
- [x] [Fix springboarding and warehouse](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872557) [[denihs](https://github.com/denihs)] [[PR](https://github.com/meteor/meteor/pull/13032)]
  - We need to fix springboarding before releasing a beta. These are tested at `tools/tests/update-tests.js`.
- [x]  https://github.com/meteor/meteor/issues/12950 [[nachocodoner](https://github.com/nachocodoner)]
- [x]  Symlink error when installing Meteor with NPM [[nachocodoner](https://github.com/nachocodoner)]
    - It seems that when installing Meteor 3 with npm, we get some errors related to symlink

</details>

<details>
  <summary><b>Meteor 3.0 RC (Started)</b></summary>

- [x] ⛏️👷 Remove any and all references to fibers throughout the code base. [[denihs](https://github.com/denihs)]
- [x] Uncaught Promise Rejection when throwing stub error [[nachocodoner](https://github.com/nachocodoner)] [[Issue](https://github.com/meteor/meteor/issues/12950)] [[PR](https://github.com/meteor/meteor/pull/13010)] 
- [x] Support Linux ARM with the Meteor installer [[nachocodoner](https://github.com/nachocodoner)] [[PR](https://github.com/meteor/meteor/pull/13077)]
- [ ] ⛏️👷 Did not check() all arguments Error [[nachocodoner](https://github.com/nachocodoner)] [[Forum Post](https://forums.meteor.com/t/meteor-3-0-beta-7-is-out/61417/3)]
- [ ] Log use of deprecated Mongo methods only for App uses [[Forum Post](https://forums.meteor.com/t/state-of-accounts-base-wrt-async/60005)]
- [ ] Remove the pinned version of packages inside the skeletons.
  - [ ] Remove `react-meteor-data` pinned versions.
      - We had to pin this version because, as react-meteor-data is not a core package, it'll always get the latest version (v2 right now) when creating a new app
  - [ ] Remove `jorgenvatle:vite-bundler` pinned versions.
- [ ] Update NPM, Node and TypeScript to latest versions [[nachocodoner](https://github.com/nachocodoner)] [[PR](https://github.com/meteor/meteor/pull/13090)]
- [ ] Circle CI tests failing for external packages [Group 5]
- [x] Sequential packages loading order [[radekmie](https://github.com/radekmie)] [[PR](https://github.com/meteor/meteor/pull/13095)]
- [ ] Using module-resolver plugin in babel throw false-positive [[issue]](https://github.com/meteor/meteor/issues/13060)

</details>

<details>
  <summary><b>Meteor 3.0 Official (To-Do)</b></summary>

<b>We need your help testing and providing feedback once the RC (Release Candidate) is out.</b>

</details>

<details>
  <summary><b>Meteor 3.x (To Do)</b></summary>

- [ ] [Bring / Update community packages](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872581)
  We want to bring some community packages to the core: meteor-migrations, meteor-synced-cron, and meteor-collection-hooks.
- [ ] Check how we could enable native ES modules on Meteor. [Context](https://github.com/meteor/meteor/discussions/12865#discussioncomment-8032938)

</details>

---

<details>
  <summary><b>Blaze (Finished)</b></summary>

- [x] Decide if any changes are necessary for the Async States in Blaze (remove Experimental warning from docs). [[radekmie](https://github.com/radekmie)] [[PR](https://github.com/meteor/blaze/pull/445)]
- [x] Discuss and decide what makes more sense within Async State Sync.
- [x] Check `firstRunPromise` [discussion](https://github.com/meteor/meteor/pull/12824). [[radekmie](https://github.com/radekmie)] [[PR](https://github.com/meteor/meteor/pull/12824)]
- [x] Support for async attributes `<div class="static {{dynamic}}" />`. [[radekmie](https://github.com/radekmie)] [[PR](https://github.com/meteor/blaze/pull/428)]
- [x] Support for async attribute spreads `<div {{dynamic}} />`. [[radekmie](https://github.com/radekmie)] [[tracking issue](https://github.com/meteor/blaze/pull/443)] [[PR](https://github.com/meteor/blaze/pull/460)]
- [x] Measure the performance impact of all async-related changes. [[radekmie](https://github.com/radekmie)] [[tracking issue](https://github.com/meteor/blaze/pull/444)]

</details>

<details>
  <summary><b>Meteor 3.0 Documentation (Started)</b></summary>

- [x] ⛏️👷 Update Meteor API Docs to v3 and migrate to Vitepress [[Grubba27](https://github.com/Grubba27)]
  - [https://v3-docs.meteor.com](https://v3-docs.meteor.com) 
- [ ] Guides
  - [ ] ⛏️👷 How to migrate to Meteor Async in Meteor 2.x
    - We are working on our own [version](https://guide.meteor.com/prepare-meteor-3.0), but led by Daniel Dornhardt, the community is also working on one. You can check the current document [here](https://docs.google.com/document/d/1XxHE5MQaS0-85HQ-bkiXxmGlYi41ggkX3F-9Rjb9HhE/edit#heading=h.65xi3waq9bb) and also the [post](https://forums.meteor.com/t/additional-async-transition-documentation/60739) in the forum.
  - [ ] How to migrate from Meteor 2.x to Meteor 3.0
- [ ] Update Tutorials
  - [ ] Update Meteor 3.0 with React Tutorial
  - [ ] Update Meteor 3.0 with Blaze Tutorial
</details>

<details>
  <summary><b>Independent releases (Started)</b></summary>

- [x] Update Meteor 3.0 to use Reify 0.25.1
    - Some places are still using version 0.24.0
- [ ] Review compiler-plugin addStylesheet.
  - We have a partial fix for stylesheets that we need to make if we want to keep them. For more [context](https://github.com/meteor/meteor/pull/12538).
- [ ] [CoffeeScript](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872760)
  - Test everything related to CoffeeScript and make sure it works.

</details>

<details>
  <summary><b>Performance / Helpers (To Do)</b></summary>

- [ ] [Email package](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=27872785)
  - Analyze if it's possible and/or if it has good performance even using TLA.
- [ ] [Improve HMR performance](https://github.com/orgs/meteor/projects/10?pane=issue&itemId=33121116)
  - Make sure the changes we did with HMR don't hurt the performance. For more [context](https://github.com/meteor/meteor/pull/12542).
- [ ] Create a stress test to compare data between Meteor 2 and Meteor 3, now without Fibers and Node 20.
  - We want to create materials comparing Meteor 2 and 3 performance.
- [ ] Evaluate how we can create tools to help in the migration process.
  - The community has created a [Codemod](https://forums.meteor.com/t/how-did-i-migrate-our-app-to-new-async-api-meteor-2-8/59295), Meteor [packages](https://dev.to/jankapunkt/prepare-your-meteorjs-project-for-the-big-30-release-14bf) and [ESLint plugins](https://forums.meteor.com/t/meteor-3-0-async-functions-eslint-plugin/60738/19).

</details>

> ⛏️👷 means someone is working on it

**Comment by markpdev2** [comment]
Hi just checking ETA  of 3.0 beta -- thanks heaps

**Reply by rj-david** [reply]
https://forums.meteor.com/t/whats-left-until-an-official-meteor-3-0/60775

**Comment by trusktr** [comment]
I see `accounts-password` on the list, but I think `accounts-ui` needs to be on the list too, as that had a constraint error when I tried to upgrade to 3.0 beta.0.

**Comment by trusktr** [comment]
It would be really great if native ES module (ESM) support is slated for 3.0. Meteor is already lagging behind, and if 3.0 comes out without native ES module support, it'll just still be lagging behind in a fairly critical way: the `ecmascript` package compiles all code to an internal CommonJS format, and has a chance of being incompatible with today's libraries if they rely on things like Node.js `exports`.

I tried to add `type:module` to my app's `package.json` and it causes errors in Meteor 3.0 beta.0 like so:

https://forums.meteor.com/t/meteor-3-0-type-module-in-package-json-causes-error/61084

I also tried find a workaround to import native Node ES modules on the _server side_, but have not yet been successful:

https://forums.meteor.com/t/unable-to-import-native-es-modules-in-meteor-2-0-or-3-0-beta-0/61085

(it is easy to run client code as vanilla JS modules, that post links to an example)

**Reply by StorytellerCZ** [reply]
Agreed, this is a must for me as well as I have packages that I need to upgrade with use the Node.js `exports`.

**Comment by harryadel** [comment]
MS folks @fredmaiaarantes @Grubba27 

I believe this discussion is long overdue for some updates on your part. Looking forward to it. :+1:

**Reply by Grubba27** [reply]
Hey Harry! I'm currently working on migrating docs(and updating them to be correct with the API in v3.0). You can check the PR [here](https://github.com/meteor/meteor/pull/12926) (there is a todo list of what I'm doing and what is left)

Also, @denihs [posted](https://github.com/meteor/meteor/discussions/11505#discussioncomment-8100901) last week about our progress

**Reply by harryadel** [reply]
Thank you for clarifying things up. It's just we're kept in the dark and get a random update once every while about previously done work but no future plans. You're working on updating the docs and denis and on source map error, is that it? 

>⛏️👷 means someone is working on it

>⛏️👷 Update Meteor API Docs to v3 and migrate to Vitepress [[Grubba27](https://github.com/Grubba27)]

>⛏️👷 Error with Source-map showing up in some cases. [Context](https://meteor-community.slack.com/archives/C01SWPPE81F/p1702314166182529). [[denihs](https://github.com/denihs)]

**Reply by dallman2** [reply]
@Grubba27 @harryadel @fredmaiaarantes I have looked through the backlog and the Meteor beta code, as well as the progress posts referenced above. However, I cannot really find an authoritative answer on the plan for `call`/ `apply`. Will these methods be **deprecated** in favor of `callAsync`/ `applyAsync`? Will they still exist in the first release of Meteor 3? As of `3.0-beta.0`, they are still there and functioning.

Thanks guys

**Reply by denihs** [reply]
Hi @dallman2, in version 3, you'll still be able to use these methods normally. We don't have plans to remove them.

They will still work as they're working now in version 2, but of course, there are scenarios where you don't want to use them. For example, when calling async Methods, especially the ones that handle async stubs (Mongo methods).

**Comment by ramesh-bhatta-oohmedia** [comment]
Hello Meteor Community,

I hope this message finds you well. I am reaching out to seek assistance regarding issues I encountered after updating my Meteor project to version 3.0-beta.6.
Here's a brief overview of the situation:
I created my app using the command: meteor create --vue simple-todos-vue.
Following that, I decided to update my project to Meteor version 3.0-beta.6 using the command: 
meteor update --release 3.0-beta.6.
However, post-update, I am facing a specific issue related to the 'jorgenvatle:vite-bundler' package. The error message is as follows:
javascriptCopy codeWhile loading plugin `vite` from package `jorgenvatle:vite-bundler`:
packages/modules-runtime.js:222:12: Cannot find module 'fibers'

I have attempted to troubleshoot this issue by consulting the official Meteor documentation and various online resources. Unfortunately, my efforts have not led to a resolution.

If anyone in the community has encountered a similar problem or has insights into potential solutions for the 'Cannot find module 'fibers'' issue, I would greatly appreciate your guidance.

To provide more context, I have also checked the package versions and ensured that 'fibers' is installed in my project using meteor npm install fibers. Despite these efforts, the issue persists.

If you require additional details or specific error logs, please let me know, and I will promptly provide them.
Thank you in advance for your time and support. I look forward to any insights or suggestions the community can offer.

Thanks,
Ramesh

**Reply by StorytellerCZ** [reply]
Hi @ramesh-bhatta-oohmedia, there is already an issue for this: https://github.com/meteor/meteor/issues/13057

**Comment by dallman2** [comment]
Hey all, 

Are there any new version 3 betas in the pipeline? I have been seeing a lot of activity on the draft PR, but didn't catch another beta version. Should we be expecting an RC next? I am getting a little worried about the deadline for the end of security patches on Node 14 looming...

**Reply by denihs** [reply]
Hi, just wrote a [quick update](https://forums.meteor.com/t/fibers-public-roadmap-and-meteor-3-0/59627/103?u=denyhs) in our Forums about this.

We plan to release our last Beta between this week and the next. After that, we'll start working in our first RC.

**Comment by fredmaiaarantes** [comment]
I organized the [discussion](https://github.com/meteor/meteor/discussions/12865?sort=new#discussion-5820005) and added estimated release dates for the last Beta and first RC versions.

**Reply by dallman2** [reply]
Thanks for adding that. This makes it much easier for us to plan our migration timeline.

**Reply by harryadel** [reply]
Thanks for clearing things up. :+1:

**Comment by dhanalakshmi3467** [comment]
account-base ui npm add throwing   errors

I am currently using Meteor js version 2.15 and need to install accounts-base-ui. However, when attempting to add wiseguyeh:accounts-azure-active-directory, I encounter errors. Could you please advise on the appropriate version?

**Reply by StorytellerCZ** [reply]
Please check with the package maintainer. Given that the package had last update 10 years ago, I would recommend finding a newer version of Azure OAuth. It is unrelated to this discussion topic.

**Comment by henriquealbert** [comment]
Updates about the first Meteor 3.0 RC Release:

Due to some roadblocks, our ETA for the RC Release has been postponed to **19th April.**
These are the last tasks that we are finishing:

- Sequential packages loading order [[radekmie](https://github.com/radekmie)] [[PR](https://github.com/meteor/meteor/pull/13095)]
- Update NPM, Node and TypeScript to latest versions [[nachocodoner](https://github.com/nachocodoner)] [[PR](https://github.com/meteor/meteor/pull/13090)] - a nice to have but not a blocker.

**Comment by henriquealbert** [comment]
It is time for a Weekly update :tada: :comet: 

We are proud to say that we have the **first Release Candidate for Meteor 3.0** 🤩🥳
You can check more details about it [here](https://forums.meteor.com/t/meteor-3-0-rc-0-is-out/61515).

The priority of the next RC will be to review and fix issues with Cordova's experience.

### Next Releases:
- Meteor 2.16 [[PR]](https://github.com/meteor/meteor/pull/13066)
  - We plan to release it by the end of this week. 
- Meteor 3.0 RC (next): May 3, 2024. 
  - The plan is to have a new release candidate every 2 weeks.

---
You can follow the updates on [What's left until an official Meteor 3.0? Discussion](https://github.com/meteor/meteor/discussions/12865), and on the [Release 3.0 PR](https://github.com/meteor/meteor/pull/12359).

**Comment by henriquealbert** [comment]
It is time for a Weekly update :tada: :comet: 

We are proud to say that we have the **Release for 2.16 beta** :star_struck::partying_face:
You can check more details about it [here ](https://forums.meteor.com/t/meteor-v2-16-first-beta-is-out/61546).


Tasks and fixes for the next Meteor 3.0 RC:
- Circle CI tests failing for external packages [Group 5]
- [Meteor developer accounts do not work when logging in with github · Issue #12418](https://github.com/meteor/meteor/issues/12418) 
- [Add missing observe async types by StorytellerCZ · Pull Request #13103](https://github.com/meteor/meteor/pull/13103)
- [Added `waitUntilAllLoaded` safety check. by radekmie · Pull Request #13102](https://github.com/meteor/meteor/pull/13102)
- [refactor(accounts): add ddp type to accounts-client by vit0rr · Pull Request #13042](https://github.com/meteor/meteor/pull/13042)
- Make sure Vue and Svelte apps run with Meteor 3
 

### Next Releases:
- Meteor 2.16 RC [[PR]](https://github.com/meteor/meteor/pull/13066)
  - We plan to release it early next week, probably Monday (May 6th).
- Meteor 3.0 RC (next): Friday, May 3rd, 2024. 
  - The plan is to have a new release candidate every 2 weeks.

---
You can follow the updates on [What's left until an official Meteor 3.0? Discussion](https://github.com/meteor/meteor/discussions/12865), and on the [Release 3.0 PR](https://github.com/meteor/meteor/pull/12359).

**Comment by henriquealbert** [comment]
It is time for a Weekly update :tada: :comet: 

We are proud to say that we have the **Release for Meteor 3.0-rc.1** :star_struck::partying_face:
You can check more details about it [here](https://forums.meteor.com/t/meteor-3-0-rc-1-is-out/61580).

We also have the **Meteor 2.16-rc.0** :star_struck::partying_face:
You can check more details about it [here](https://forums.meteor.com/t/meteor-v2-16-first-rc-is-out/61582).


---


Tasks and fixes for the **next** 3.0 RC:
- Review the Types for WebApp/Express
- [[Meteor 3] Fix handling unexpected errors during build by zodern · Pull Request #12675](https://github.com/meteor/meteor/pull/12675)
- Fix and Review any places using findOne in the server
- [[Meteor 3] Environment Variables resets value too early · Issue #13112](https://github.com/meteor/meteor/issues/13112)
- [[Meteor 3] EnvironmentVariable.withValue has different return value on client and server · Issue #13113](https://github.com/meteor/meteor/issues/13113)
- [Unable to create new meteor 3 projects · Issue #13106](https://github.com/meteor/meteor/issues/13106)
- [Make sure apollo skeleton works with Meteor 3](https://forums.meteor.com/t/meteor-create-apollo-not-yet-compatible-with-meteor-3-0/61545)
- Docs for all modules strict breaking change
- Review security updates
 

### Next Releases:
- **Meteor 2.16** - Official [[PR]](https://github.com/meteor/meteor/pull/13066)
  - We plan to release it next week. Probably until May 14th.
- **Meteor 3.0 RC** - next [[PR]](https://github.com/meteor/meteor/pull/12359)
  - Monday, May 20th, 2024. 
  - The plan is to have a new release candidate every 2 weeks. Every Monday.
- **Meteor 3.0** - Official 
  - We need your help testing the RC versions to make it more stable before the official release.

**Comment by henriquealbert** [comment]
It is time for a Weekly update :tada: :comet: 

We are proud to say that we have the **Release for Meteor 2.16 official** :star_struck::partying_face:
You can check more details about it [here](https://forums.meteor.com/t/meteor-2-16-is-out/61612).

---

### Next Releases:
- **Meteor 3.0 RC** - next [[PR]](https://github.com/meteor/meteor/pull/12359)
  - Monday, May 20th, 2024. 
  - The plan is to have a new release candidate every 2 weeks. Every Monday.
- **Meteor 3.0** - Official 
  - We need your help testing the RC versions to make it more stable before the official release.
- **Meteor 2.X versions**
  - We plan to hold off on releasing new Meteor 2.X versions for now to focus on the 3.0 releases.

**Comment by leonardoventurini** [comment]
It's time for a weekly update :tada: :comet:
It's been some time, but we are proud to say that we have the Release for Meteor 3.0-rc.2 :star_struck:  :partying_face:

You can check [here](https://forums.meteor.com/t/meteor-3-0-rc-2-is-out/61636) for more details.

Tasks and fixes for the next, and final until the official, Meteor 3.0 Release Candidate (rc.3):

* [Missing Accounts.addEmailAsync](https://github.com/meteor/meteor/pull/13164)
* [Top-level await does not seem to work in Meteor 3](https://github.com/meteor/reify/pull/8)
* [AllowDeny async compatibility](https://github.com/meteor/meteor/pull/13168)
* [Fix WebApp.addRuntimeConfigHook](https://github.com/meteor/meteor/pull/13156)
* [Implement reset db option and default without cleaning db](https://github.com/meteor/meteor/pull/13162)
* [Upgrade meteor presets packages that are deprecated](https://github.com/meteor/meteor/pull/13171)
* Upgrade Blaze to latest version
* MultiFileCachingCompiler does not set up its internal LRU cache correctly
* Incorrect Node.js version in Meteor 3 projects
* Add doc for vite-bundler in the migration guide
* Review security updates

`Meteor 3-rc.3` is expected to arrive on Tuesday, June 11th, stay tuned.

## Next Releases

* Meteor 3.0 Official (**July 8th, 2024**) [[PR]](https://github.com/meteor/meteor/pull/13163)
* Meteor 3.0.1 (July 22nd, 2024)
* Meteor 3.0.2 (August 5th, 2024)

These dates are subject to change as we add or remove things from the backlog.

**Comment by leonardoventurini** [comment]
It's time for a weekly update :tada: :comet:

We are proud to say that we have the Release for Meteor 3.0-rc.4 :star_struck:  :partying_face:

You can check [here](https://forums.meteor.com/t/meteor-3-0-rc-4-is-out/61742) for more details.

---

Tasks and fixes for the next, and official, **Meteor 3.0 Release**:

* Ensure the documentation is up to date
* Upgrade Node.js to v20.14
* Improve package version resolution process
* React Router SSR for Meteor 3
* Remove pinned version of packages inside the skeletons
* Add `addEmailAsync` alias
* Fix incorrect Node.js version in Meteor 3 projects
* Review Security Updates

**Meteor 3.0** is expected to arrive on Tuesday, July 8th, stay tuned.

## Next Releases

* Meteor 3.0.1 (July 22nd, 2024)
* Meteor 3.0.2 (August 5th, 2024)

These dates are subject to change as we add or remove things from the backlog.

**Reply by CharlyTalavera** [reply]
Hi @leonardoventurini! Just FYI the status in the migration docs looks outdated. It still saying `3.0-rc.2` and Node version `20.12.2` (Here https://guide.meteor.com/3.0-migration).

**Reply by leonardoventurini** [reply]
Thanks for reporting it, we are going to replace that with the proper and most up-to-date migration docs app today.

**Comment by leonardoventurini** [comment]
***Weekly Update June 28th, 2024*** :tada: :comet:

So far we have released `rc4` as we mentioned in our previous updates, and work is progressing well for our official Meteor 3 release. 

You can check [here](https://forums.meteor.com/t/meteor-3-0-rc-4-is-out/61742) for more details about our last Release Candidate.

The backlog has evolved a little over the last days with some issues gaining priority, our most important ones are as follows:

- Incorrect Node.js version in Meteor project if the `meteor-tool` is missing, this has gone up in priority but it's extremely hard to solve, we have a path forward but have some loose ends we need to account for, some tests on Windows are failing at the moment
- The method call order when Meteor goes back from offline and authentication
- Finding a replacement for `@vlasky/whomst` since it depends on a package called `posix` which builds a binary and breaks on some installations

We also confirmed that Cordova is working completely fine, this was done by @nachocodoner, one of our developers with the most expertise in that area.

We are also revamping the documentation and researching ways to simplify the mechanics for `call` and  `callAsync`, we have @denyhs at this forefront and also huge help from @zodern . The goal is to make Meteor have the best Developer Experience as possible and the migration as simple as possible too.

---

Other tasks and fixes for the next, and official, **Meteor 3.0 Release**:

- Unable to run bundle from "meteor build"
- Ensure the documentation is up to date
- Upgrade Node.js to v20.14
- Improve package version resolution process
- React Router SSR for Meteor 3
- Remove pinned version of packages inside the skeletons
- Add `addEmailAsync` alias
- Incorrect Node.js version in Meteor 3 projects
- Review Security Updates
- Rename or remove `asyncLocalStorage` global
- Verify `tinytest` major version increase

**Meteor 3.0** is expected to arrive on Monday, July 8th or 15th at the most

**Next Releases**

* Meteor 3.0.1 (July 22nd, 2024)
* Meteor 3.0.2 (August 5th, 2024)

These dates are subject to change as we add or remove things from the backlog.

**Reply by StorytellerCZ** [reply]
Please account for this before the final release:
https://nodejs.org/en/blog/vulnerability/july-2024-security-releases

**Reply by vlasky** [reply]
> * Finding a replacement for `@vlasky/whomst` since it depends on a package called `posix` which builds a binary and breaks on some installations

@leonardoventurini I'm the author of [@vlasky/whomst](https://github.com/vlasky/whomst). Are there any changes I can make to help resolve the issue?

I see other potential replacements for the posix NPM dependency, e.g. [posix-node](https://www.npmjs.com/package/posix-node) and [libsys](https://github.com/streamich/libsys). Would they experience the same build issue?

I currently do my building in WSL2 Ubuntu under Windows 10 and in Rocky Linux 9.

**Reply by vlasky** [reply]
@leonardoventurini I can also fork the original posix NPM package and make fixes to it.

**Reply by leonardoventurini** [reply]
Ideally we don't want to have dependencies that require building and things like that, @nachocodoner is working on this.

**Reply by nachocodoner** [reply]
Hello @vlasky!

We are addressing this issue in [this PR](https://github.com/meteor/meteor/pull/13204) by replacing the [@vlasky/whomst](https://github.com/vlasky/whomst) dependency with a custom implementation that doesn't rely on a native npm package like posix or other. The main problem is that posix fails to be properly included as an optional npm dependency with native requirements on certain OS, such as Windows. 

This issue is not about the [whomst](https://github.com/vlasky/whomst) library itself I think, but rather the requirement of a native package and the Meteor package resolver, which struggles on prepare specific OS dependencies in certain circumstances. We have identified a possible replacement for this library in our specific use case to simplify the fix and as we don't use intensively whomst features.

We only need to obtain the groupId from a groupName, [something that can be done by running a few commands on any UNIX system](https://github.com/meteor/meteor/pull/13204/files#diff-95c25d104865a5165f4b922122c8ad2ca9352ab1368613e863c608c4aaac7214R1483-R1508). This is required only for a specific feature in Meteor, activated when using the `UNIX_SOCKET_GROUP` environment variable, and needing a UNIX environment.

**Comment by leonardoventurini** [comment]
***Weekly Update July 5th, 2024*** :tada: :comet:

So far we have released `rc4` as we mentioned in our previous updates, and work continues progressing well for our Meteor 3 official release.

You can check [here](https://forums.meteor.com/t/meteor-3-0-rc-4-is-out/61742) for more details about our last Release Candidate.

We have launched our **Meteor Lounge** Discord server, you read more about it [here](https://forums.meteor.com/t/welcome-to-the-meteor-lounge-our-new-official-discord-server/61821).

Today we have successfully hosted our first Community Call where many amazing people joined like  @StorytellerCZ @copleykj , to name just a few! We are going to host these regularly, the time is still to be defined. If you have any suggestions please reply below.

[Join Meteor Lounge](https://discord.gg/hZkTCaVjmT)

---

Tasks and fixes for the next, and official, **Meteor 3.0 Release**:

- Incorrect Node.js version in Meteor 3 projects
- Find alternative and remove dependencies which hold the `posix` package
- Check Fix handling unexpected errors during build and merge it
- Unable to run bundle from "meteor build"
- Deprecate Meteor.call for async methods
- Review and update documentation for `callAsync`
- Make sure Cordova is working well (native build)
- Replace docs page for Meteor 3 with a link to the ad hoc docs for the migration
- Fix how to compare constraints version when publishing a package
- Check if TLA or require as promises is missing documentation
- Update node version to latest v20
- Rename or remove asyncLocalStorage global
- Tinytest major version increased
- Add `addEmailAsync` alias
- Cannot create new project with `meteor create my-react-app` due to skeleton clone

**Meteor 3.0** is expected to arrive from Monday, July 8th to July 15th at the most. Most likely sometime mid-week.

**Next Releases**

* Meteor 3.0.1 (July 29th, 2024)
* Meteor 3.0.2 (August 12th, 2024)

These dates are subject to change as we add or remove things from the backlog.

**Reply by StorytellerCZ** [reply]
Given this: https://nodejs.org/en/blog/vulnerability/july-2024-security-releases
I would be fine if it is delayed till later in the week, so that it can be included.
Also I think this needs to be addressed before release as well: https://github.com/meteor/meteor/issues/13212

**Comment by cbnsndwch** [comment]
Now that 3.0 is out, should this discussion be closed or renamed?

https://dev.to/meteor/meteor-3-is-officially-here-3gml

**Reply by fredmaiaarantes** [reply]
Yes! 🎉
